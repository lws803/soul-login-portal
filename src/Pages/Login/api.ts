import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const loginWithPlatform = async ({
  platformId,
  callback,
  values,
  state,
  codeChallenge,
}: {
  platformId: number;
  callback: string;
  values: {
    email: string;
    password: string;
  };
  state: string;
  codeChallenge: string;
}): Promise<{
  data: LoginWithPlatformResponse | null;
  error: ApiError | null;
}> => {
  try {
    const { data } = await axios.post<LoginWithPlatformResponse>(
      `${BASE_URL}/v1/auth/code`,
      values,
      {
        params: {
          client_id: platformId,
          redirect_uri: callback,
          state,
          code_challenge: codeChallenge,
        },
      },
    );
    return { data, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};

type LoginWithPlatformResponse = {
  code: string;
  state: string;
};

export const login = async ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => {
  try {
    const { data } = await axios.post<LoginResponse>(
      `${BASE_URL}/v1/auth/login`,
      { email, password },
    );
    return { data, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};

type LoginResponse = {
  access_token: string;
};

export const joinPlatformAndLogin = async ({
  platformId,
  callback,
  email,
  password,
  accessToken,
  state,
  codeChallenge,
}: {
  platformId: number;
  callback: string;
  email: string;
  password: string;
  accessToken: string;
  state: string;
  codeChallenge: string;
}) => {
  try {
    await axios.post(`${BASE_URL}/v1/platforms/${platformId}/join`, undefined, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });

    const { data: codeData } = await axios.post<LoginWithPlatformResponse>(
      `${BASE_URL}/v1/auth/code`,
      { email, password },
      {
        params: {
          platform_id: platformId,
          redirect_uri: callback,
          state,
          code_challenge: codeChallenge,
        },
      },
    );
    return { data: codeData, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};
