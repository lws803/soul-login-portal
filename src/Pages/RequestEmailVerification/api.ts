import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const sendEmailConfirmation = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/resend-confirmation-token`,
      {},
      { params: { email } },
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
