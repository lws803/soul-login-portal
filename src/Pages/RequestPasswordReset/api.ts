import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const sendPasswordReset = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/request-password-reset-token`,
      {},
      { params: { email } },
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
