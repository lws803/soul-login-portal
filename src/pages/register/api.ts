import axios from 'axios';

import { BASE_URL } from '../../constants';
import { ApiError } from '../../shared/api-types';

export const register = async ({
  values,
}: {
  values: { email: string; password: string; username: string };
}) => {
  try {
    const { data } = await axios.post<Response>(`${BASE_URL}/v1/users`, values);
    return { data, error: null };
  } catch ({ response: { data, status } }) {
    return {
      data: null,
      error: { data: data as ApiError, status: status as number },
    };
  }
};

type Response = {
  id?: number;
};
