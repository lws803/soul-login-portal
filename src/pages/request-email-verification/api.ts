import axios from 'axios';

import { BASE_URL } from '../../constants';
import { ApiError } from '../../shared/api-types';

export const sendPasswordReset = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/request-password-reset-token?email=${email}`,
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
