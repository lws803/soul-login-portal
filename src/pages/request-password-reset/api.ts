import axios from 'axios';

import { BASE_URL } from '../../constants';

import { ApiError } from '../../shared/api-types';

export const sendEmailConfirmation = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/resend-confirmation-token?email=${email}`,
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
