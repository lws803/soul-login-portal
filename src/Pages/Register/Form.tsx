import { useState } from 'react';
import {
  FormControl,
  FormLabel,
  Input,
  FormHelperText,
  FormErrorMessage,
  Box,
  Progress,
} from '@chakra-ui/react';
import * as Yup from 'yup';
import zxcvbn from 'zxcvbn';
import { useFormik } from 'formik';
import { useNavigate, useLocation } from 'react-router-dom';

import FancyButton from 'src/components/FancyButton';

import { register } from './api';

export default function Form({ setErrors }: Props) {
  const navigate = useNavigate();
  const { search } = useLocation();

  const [isRegistering, setIsRegistering] = useState(false);

  const formik = useFormik({
    initialValues: {
      email: '',
      password: '',
      username: '',
    },
    onSubmit: async (values) => {
      setErrors([]);
      setIsRegistering(true);

      const { data, error } = await register({ values });
      if (error) {
        setIsRegistering(false);
        if (error.status === 400 && error.data.error === 'VALIDATION_ERROR') {
          formik.setErrors({
            password: 'Password is too weak!',
          });
        } else if (error.status === 409) {
          formik.setErrors({ email: error.data.message });
        } else {
          setErrors([error.data.message]);
        }
      }
      if (data?.id) {
        navigate(`/${search}`);
      }
    },
    validationSchema: Yup.object({
      email: Yup.string().email().required(),
      username: Yup.string().required(),
      password: Yup.string().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.username && formik.touched.username}
      >
        <FormLabel htmlFor="username">Username</FormLabel>
        <Input
          id="username"
          name="username"
          onChange={formik.handleChange}
          value={formik.values.username}
          aria-label="Username input"
          variant="filled"
          placeholder="Your username"
          disabled={isRegistering}
        />
        {!formik.errors.email && (
          <FormHelperText>Choose an awesome username!</FormHelperText>
        )}
        {formik.errors.username && (
          <FormErrorMessage>{formik.errors.username}</FormErrorMessage>
        )}
      </FormControl>

      <FormControl
        isInvalid={!!formik.errors.email && formik.touched.email}
        marginTop={8}
      >
        <FormLabel htmlFor="email">Email</FormLabel>
        <Input
          id="email"
          name="email"
          onChange={formik.handleChange}
          value={formik.values.email}
          aria-label="Email input"
          variant="filled"
          placeholder="Your email"
          disabled={isRegistering}
        />
        {!formik.errors.email && (
          <FormHelperText>We&apos;ll never share your email.</FormHelperText>
        )}
        {formik.errors.email && (
          <FormErrorMessage>{formik.errors.email}</FormErrorMessage>
        )}
      </FormControl>
      <FormControl
        isInvalid={!!formik.errors.password && formik.touched.password}
        marginTop={8}
      >
        <FormLabel htmlFor="password">Password</FormLabel>
        {formik.values.password.length > 0 && (
          <Progress
            value={((zxcvbn(formik.values.password).score + 1) / 5) * 100}
            colorScheme="soul.green"
            size="xs"
            mb={1}
          />
        )}
        <Input
          id="password"
          name="password"
          type="password"
          onChange={formik.handleChange}
          value={formik.values.password}
          aria-label="Password input"
          variant="filled"
          placeholder="Your password"
          disabled={isRegistering}
        />
        {formik.errors.password && (
          <FormErrorMessage>{formik.errors.password}</FormErrorMessage>
        )}
      </FormControl>
      <Box mt={8}>
        <FancyButton
          isLoading={isRegistering}
          disabled={isRegistering}
          type="submit"
        >
          Register
        </FancyButton>
      </Box>
    </form>
  );
}

type Props = {
  setErrors: (errors: string[]) => void;
};
