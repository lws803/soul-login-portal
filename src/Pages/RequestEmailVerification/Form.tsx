import { useFormik } from 'formik';
import * as Yup from 'yup';
import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  HStack,
  Button,
} from '@chakra-ui/react';

import { sendEmailConfirmation } from './api';

export default function Form({ setErrors, setIsSuccess }: Props) {
  const formik = useFormik({
    initialValues: {
      email: '',
    },
    onSubmit: async (values) => {
      setErrors([]);
      const { error } = await sendEmailConfirmation(values);

      if (error) {
        setErrors([error.message]);
      } else {
        setIsSuccess(true);
      }
    },
    validationSchema: Yup.object({
      email: Yup.string().email().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.email && formik.touched.email}
        marginTop={8}
      >
        <FormLabel htmlFor="email">Email</FormLabel>
        <Input
          id="email"
          name="email"
          onChange={formik.handleChange}
          value={formik.values.email}
          aria-label="Email input"
          variant="filled"
          placeholder="Enter your email"
        />
        {formik.errors.email && (
          <FormErrorMessage>{formik.errors.email}</FormErrorMessage>
        )}
      </FormControl>
      <HStack mt={8}>
        <Button
          type="submit"
          isLoading={formik.isSubmitting}
          loadingText="Sending verification email"
          width="100%"
          bg="soul.pink.light"
          _hover={{ bg: 'soul.pink.dark' }}
          _active={{ bg: 'soul.pink.dark' }}
        >
          Request email verification
        </Button>
      </HStack>
    </form>
  );
}

type Props = {
  setErrors: (errors: string[]) => void;
  setIsSuccess(isSuccess: boolean): void;
};
