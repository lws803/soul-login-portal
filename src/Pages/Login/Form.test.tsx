import React from 'react';
import { fireEvent, render, act, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import Form from './Form';
import * as api from './api';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
  useLocation: () => ({ search: '?SEARCH' }),
}));

describe(Form, () => {
  const props: React.ComponentProps<typeof Form> = {
    setErrors: jest.fn(),
    platformId: 1,
    callback: 'https://www.example.com',
    disabled: false,
    state: 'STATE',
    codeChallenge: 'CODE_CHALLENGE',
    setJoinPlatform: jest.fn(),
    joinPlatform: false,
  };

  it('renders', async () => {
    const { container } = render(
      <MemoryRouter>
        <Form {...props} />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('logs in successfully', async () => {
    window.open = jest.fn();

    const loginWithPlatform = jest
      .spyOn(api, 'loginWithPlatform')
      .mockResolvedValue({
        data: { code: 'CODE', state: 'STATE' },
        error: null,
      });

    const { getByLabelText, getByText } = render(
      <MemoryRouter>
        <Form {...props} />
      </MemoryRouter>,
    );

    act(() => {
      fireEvent.change(getByLabelText('Email input'), {
        target: { value: 'TEST@EMAIL.COM' },
      });
      fireEvent.change(getByLabelText('Password input'), {
        target: { value: 'PASSWORD' },
      });
    });
    fireEvent.click(getByText('Login'));

    await waitFor(() => {
      expect(loginWithPlatform).toHaveBeenCalledWith({
        callback: 'https://www.example.com',
        platformId: 1,
        values: {
          email: 'TEST@EMAIL.COM',
          password: 'PASSWORD',
        },
        state: 'STATE',
        codeChallenge: 'CODE_CHALLENGE',
      });
      expect(window.open).toHaveBeenCalled();
    });
  });

  it('redirects to register page', async () => {
    jest.spyOn(api, 'loginWithPlatform').mockResolvedValue({
      data: null,
      error: { error: 'USER_NOT_FOUND', message: '' },
    });

    const { getByLabelText, getByText } = render(
      <MemoryRouter>
        <Form {...props} />
      </MemoryRouter>,
    );

    act(() => {
      fireEvent.change(getByLabelText('Email input'), {
        target: { value: 'TEST@EMAIL.COM' },
      });
      fireEvent.change(getByLabelText('Password input'), {
        target: { value: 'PASSWORD' },
      });
    });
    fireEvent.click(getByText('Login'));

    await waitFor(() => {
      expect(mockedUsedNavigate).toHaveBeenCalledWith('/register?SEARCH');
    });
  });

  it('shows join platform button', async () => {
    const setJoinPlatform = jest.fn();

    jest.spyOn(api, 'loginWithPlatform').mockResolvedValue({
      data: null,
      error: { error: 'PLATFORM_USER_NOT_FOUND', message: '' },
    });
    jest.spyOn(api, 'login').mockResolvedValue({
      data: { access_token: 'ACCESS_TOKEN' },
      error: null,
    });

    const { getByLabelText, getByText } = render(
      <MemoryRouter>
        <Form {...props} setJoinPlatform={setJoinPlatform} />
      </MemoryRouter>,
    );

    act(() => {
      fireEvent.change(getByLabelText('Email input'), {
        target: { value: 'TEST@EMAIL.COM' },
      });
      fireEvent.change(getByLabelText('Password input'), {
        target: { value: 'PASSWORD' },
      });
    });
    fireEvent.click(getByText('Login'));

    await waitFor(() => {
      expect(setJoinPlatform).toHaveBeenCalledWith(true);
    });
  });
});
