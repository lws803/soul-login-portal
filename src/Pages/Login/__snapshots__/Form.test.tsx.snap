// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders 1`] = `
<div>
  <form>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="email"
        id="field-1-label"
      >
        Email
      </label>
      <input
        aria-label="Email input"
        class="chakra-input css-0"
        id="email"
        name="email"
        placeholder="Your email"
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-164r41r"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="password"
        id="field-2-label"
      >
        Password
      </label>
      <input
        aria-describedby="field-2-helptext"
        aria-label="Password input"
        class="chakra-input css-0"
        id="password"
        name="password"
        placeholder="Your password"
        type="password"
        value=""
      />
      <div
        class="chakra-form__helper-text css-0"
        id="field-2-helptext"
      >
        <a
          class="chakra-link css-0"
          href="/request-password-reset"
        >
          Forgot password?
        </a>
      </div>
    </div>
    <div
      class="css-164r41r"
    >
      <button
        class="css-y6mgpv"
        type="submit"
      >
        <div
          class="css-tkevr6"
        >
          Login
        </div>
      </button>
    </div>
  </form>
  <div
    class="css-19ug521"
  >
    <a
      class="chakra-link css-0"
      href="/request-email-verification"
    >
      Resend email verification
    </a>
  </div>
</div>
`;
